tempesta_fw/vhost.c

```diff
@@ -1937,18 +2239,15 @@ tfw_cfgop_vhosts_list_free(TfwVhostList *vhosts)
 static int
 tfw_vhost_start(void)
 {
-	if (!tfw_runstate_is_reconfig()) {
-		/* Convert Frang global timeouts to jiffies for convenience */
-		frang_cfg.clnt_hdr_timeout =
-			*(unsigned int *)&frang_cfg.clnt_hdr_timeout *
-			(unsigned long)HZ;
-		frang_cfg.clnt_body_timeout =
-			*(unsigned int *)&frang_cfg.clnt_body_timeout *
-			(unsigned long)HZ;
-	}
+	TfwVhostList *vh_list;
+
+	rcu_read_lock();
+	vh_list = rcu_dereference(tfw_vhosts);
+	rcu_read_unlock();
+	rcu_assign_pointer(tfw_vhosts, tfw_vhosts_reconfig);
```

Comment:
I believe here is an issue with multiple threads calling this function. For example, if administrator starts Tempesta multiple times in parallel. Then we can get here in two or more threads, get non-NULL `vh_list`, and then try to release its contents multiple times.

The obvious solution is to wrap rcu_dereference and rcu_assign_pointer with a lock. But since this line uses `tfw_vhosts_reconfig` which contents is updated in many places, we need a larger scope lock that protect whole configuration sequence.